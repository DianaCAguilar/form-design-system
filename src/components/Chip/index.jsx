import React from 'react';
import PropTypes from 'prop-types';
import cc from 'classcat';
import DenyIcon from 'lib/icons/react/DenyIcon';
import baseElement from 'util/baseElement';

export const THEMES = ['blue', 'gray', 'outline'];
export const SIZES = ['s', 'm'];

const Chip = ({
  size = 'm',
  theme = 'gray',
  Link,
  Icon,
  isActive,
  label,
  subtitle,
  onClose,
  ...rest
}) => {
  const Element = baseElement({ href: rest.href, onClick: true, as: Link });

  const rootClass = cc([
    {
      'fdsChip--active': isActive,
      'fdsChip--blue': theme === 'blue',
      'fdsChip--outline': theme === 'outline',
    },
    'fdsChip',
    'border--focus--noTransition',
    'transition--default',
    `fdsChip--${size}`,
  ]);
  return (
    <Element {...rest} className={rootClass}>
      {Icon && (
        <div className="margin--right--xs">
          <Icon customSize={18} />
        </div>
      )}
      <span className="fdsChip-label">
        {label}
        {subtitle && <span className="fdsChip-subtitle">{subtitle}</span>}
      </span>
      {onClose && (
        <span
          role="button"
          tabIndex="-1"
          className="fdsChip-close border--focus margin--left--xs"
          onClick={(e) => {
            onClose(e);
            e.stopPropagation();
            e.preventDefault();
          }}
        >
          <DenyIcon customSize={10} />
        </span>
      )}
    </Element>
  );
};

Chip.propTypes = {
  /**
   * Text that appears in the button
   */
  label: PropTypes.string.isRequired,
  /** Controls the color (look and feel) of the chip */
  theme: PropTypes.oneOf(THEMES),
  /** Controls the isActive state of the chip, which changes colors */
  /** Specify the size of the chip */
  size: PropTypes.oneOf(SIZES),
  isActive: PropTypes.bool,
  /**  Pass in "only" a FDS Icon reference to display it (e.g. Icon={ApproveIcon}) */
  Icon: PropTypes.func,
  /** Text that appears to the right of the label */
  subtitle: PropTypes.string,
  /**
   * onClick handler for close icon  (By passing this onClick handler, the close icon
   * will automatically appear)
   */
  onClose: PropTypes.func,
  /**
   * Takes in a react-router `Link` reference and sets it
   * as the base element. You may ONLY use it like the
   * following:
   *
   * - `import { Link } from 'react-router'`
   * - `Link={Link}`
   */
  Link: PropTypes.func,
};

export default Chip;
