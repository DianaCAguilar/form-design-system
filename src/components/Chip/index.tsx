import React, { HTMLAttributes } from 'react';
import cc from 'classcat';
import DenyIcon from 'lib/icons/react/DenyIcon';
import baseElement from 'util/baseElement';
import Icon from 'components/Icon';

export const THEMES = ['blue', 'gray', 'outline'] as const;
export const SIZES = ['s', 'm'] as const;

export interface ChipProps
  extends HTMLAttributes<HTMLElement | HTMLButtonElement | HTMLAnchorElement> {
  /**
   * Text that appears in the button
   */
  label: string;

  /** Controls the color (look and feel) of the chip */
  theme?: typeof THEMES[number];

  /** Specify the size of the chip */
  size?: typeof SIZES[number];

  /** Controls the isActive state of the chip, which changes colors */
  isActive?: boolean;

  /**  Pass in "only" a FDS Icon reference to display it (e.g. Icon={ApproveIcon}) */
  Icon?: Icon;

  /** Text that appears to the right of the label */
  subtitle?: string;

  /**
   * onClick handler for close icon  (By passing this onClick handler, the close icon
   * will automatically appear)
   */
  onClose?: React.ReactEventHandler;

  /**
   * Takes in a react-router `Link` reference and sets it
   * as the base element. You may ONLY use it like the
   * following:
   *
   * - `import { Link } from 'react-router'`
   * - `Link={Link}`
   */
  Link?: React.ComponentType;

  href?: string;
}

const Chip = ({
  size = 'm',
  theme = 'gray',
  Link,
  Icon,
  isActive,
  label,
  subtitle,
  onClose,
  ...rest
}: ChipProps): JSX.Element => {
  const Element = baseElement({ href: rest.href, onClick: true, as: Link });

  const rootClass = cc([
    {
      'fdsChip--active': isActive,
    },
    'fdsChip',
    'border--focus--noTransition',
    'transition--default',
    `fdsChip--${size}`,
    `fdsChip--${theme}`,
  ]);
  return (
    <Element {...rest} className={rootClass}>
      {Icon && (
        <div className="fdsChip-icon">
          <Icon customSize={size === 'm' ? 16 : 14} />
        </div>
      )}
      <span>
        <span className="fdsChip-label">{label}</span>
        {subtitle && <span className="fdsChip-subtitle">{subtitle}</span>}
      </span>
      {onClose && (
        <span
          role="button"
          tabIndex={0}
          className="fdsChip-close border--focus"
          onClick={(e) => {
            onClose(e);
            e.stopPropagation();
            e.preventDefault();
          }}
        >
          <DenyIcon customSize={10} />
        </span>
      )}
    </Element>
  );
};

export default Chip;
