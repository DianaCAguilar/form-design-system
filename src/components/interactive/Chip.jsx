import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import DenyIcon from '../../../lib/icons/react/DenyIcon';
import baseElement from '../../util/baseElement';

const Chip = ({ size, Link, theme, isActive, label, subtitle, hasClose, ...rest }) => {
  const Element = baseElement({ href: rest.href, as: Link });

  const rootClass = cx('fdsChip', {
    'fdsChip--active': isActive,
    'fdsChip--small': size === 'sm',
    'fdsChip--large': size === 'lg',
    'fdsChip--blue': theme === 'blue',
    'fdsChip--outline': theme === 'outline',
  });
  return (
    <Element {...rest} className={rootClass}>
      {label}
      {subtitle && <span className="fdsChip-subtitle">{subtitle}</span>}
      {hasClose && (
        <span className="fdsChip-close">
          <DenyIcon size="xs" />
        </span>
      )}
    </Element>
  );
};

Chip.propTypes = {
  label: PropTypes.string.isRequired,
  /**
   * Pass **only** react-router `Link` here. You may **not**
   * pass anything else here: SFC, Class Component, etc (even
   * if they use react-router `Link` underneath the hood).
   */
  Link: PropTypes.func,
  isActive: PropTypes.bool,
  subtitle: PropTypes.string,
  hasClose: PropTypes.bool,
  size: PropTypes.oneOf(['sm', 'md', 'lg']),
  theme: PropTypes.oneOf(['blue', 'gray', 'outline']),
};

Chip.defaultProps = {
  theme: 'gray',
  size: 'md',
};

export default Chip;
