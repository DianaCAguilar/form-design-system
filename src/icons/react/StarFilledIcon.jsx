import React from 'react';
import withFdsIconWrapper from './withFdsIconWrapper';

const StarFilledIcon = () => (
  <svg viewBox="0 0 24 24">
    <g fillRule="evenodd">
      <g>
        <polygon points="12 17.69474 6.4376941 20.6190173 7.5 14.4252987 3 10.0388828 9.21884705 9.13523138 12 3.5 14.7811529 9.13523138 21 10.0388828 16.5 14.4252987 17.5623059 20.6190173" />
      </g>
    </g>
  </svg>
);

export default withFdsIconWrapper(StarFilledIcon);
