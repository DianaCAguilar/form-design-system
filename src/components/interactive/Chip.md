```js
import Chip from './Chip'
import { Link } from '../util/react-router';

const chips = [
  { label: 'Lorem', href: "#lorem" },
  { label: 'Ipsum', href: "#ipsum", isActive: true, subtitle: 'active' },
  { label: 'Dolor', href: "#dolor", hasClose: true },
  { label: 'Sit', onClick: () => console.log('custom logic here'), isActive: false },
];

const chipsTwo = [
  { label: 'Lorem Ipsum' },
  { label: 'Dolor Sit' },
  { Link: Link, label: 'hey', to: '#google' },
  { label: 'Sit Amet' },
  { label: 'Consectetur', onClick: () => console.log('custom logic here') },
];

<div>
  {chips.map(chip => <Chip {...chip} key={chip.label} size="sm" />)}
  <br />
  {chips.map(chip => <Chip {...chip} key={chip.label} theme="blue" size="sm" />)}
  <br /><br />
  {chips.map(chip => <Chip {...chip} key={chip.label} />)}
  <br />
  {chips.map(chip => <Chip {...chip} key={chip.label} theme="blue" />)}
  <br /><br />
  {chipsTwo.map(chip => <Chip {...chip} key={chip.label} theme="outline" hasClose={true} />)}
  <br />
</div>
```