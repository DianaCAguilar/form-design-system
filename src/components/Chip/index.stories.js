import React from 'react';
import StarFilledIcon from 'lib/icons/react/StarFilledIcon';
import { StoryWrapper, StoryItem } from 'util/storybook-docs/StoryLayout';
import Chip, { THEMES, SIZES } from '.';

const Template = (args) => <Chip {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  label: 'Hello world',
};

export const Themes = () => (
  <StoryWrapper>
    {THEMES.map((theme, idx) => (
      <StoryItem key={idx}>
        <Chip theme={theme} label={theme} />
      </StoryItem>
    ))}
  </StoryWrapper>
);

export const isActive = () => (
  <StoryWrapper>
    {THEMES.map((theme, idx) => (
      <StoryItem key={idx}>
        <Chip theme={theme} label={theme} isActive />
      </StoryItem>
    ))}
  </StoryWrapper>
);

export const Sizes = () =>
  SIZES.map((size, idx) => (
    <div key={idx} className="margin--right--s margin--bottom--s display--inlineBlock">
      <Chip size={size} label={`Size: ${size}`} />
    </div>
  ));

export const Subtitle = Template.bind({});
Subtitle.args = {
  label: 'Chip',
  subtitle: 'subtitle',
};

export const Icon = Template.bind({});
Icon.args = {
  label: 'Chip',
  Icon: StarFilledIcon,
};

export const Close = Template.bind({});
Close.args = {
  label: 'Chip',
  onClose: () => {},
};

export default {
  component: Chip,
  title: 'components/Chip',
  parameters: {
    docs: {
      attributes: ['rest'],
      description: {
        component: 'Used to display tags, members, filter settings, and other tokens.',
      },
    },
  },
};
