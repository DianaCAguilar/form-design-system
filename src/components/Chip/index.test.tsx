import React from 'react';
import { shallow } from 'enzyme';
import StarFilledIcon from 'icons/react/StarFilledIcon';
import Chip, { ChipProps } from '.';

const renderComponent = (props: ChipProps) => shallow(<Chip {...props} />);

describe('Chip component', () => {
  it('matches snapshot (default props)', () => {
    const component = renderComponent({
      label: 'foo',
    });
    expect(component).toMatchSnapshot();
  });

  it('matches snapshot (set all props)', () => {
    const component = renderComponent({
      label: 'foo',
      isActive: true,
      subtitle: 'bar',
      onClose: () => {},
      Icon: StarFilledIcon,
      size: 'm',
      theme: 'blue',
    });
    expect(component).toMatchSnapshot();
  });
});
