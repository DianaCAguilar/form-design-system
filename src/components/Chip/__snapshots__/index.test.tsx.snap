// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip component matches snapshot (default props) 1`] = `
<button
  className="fdsChip border--focus--noTransition transition--default fdsChip--m fdsChip--gray"
>
  <span>
    <span
      className="fdsChip-label"
    >
      foo
    </span>
  </span>
</button>
`;

exports[`Chip component matches snapshot (set all props) 1`] = `
<button
  className="fdsChip--active fdsChip border--focus--noTransition transition--default fdsChip--m fdsChip--blue"
>
  <div
    className="fdsChip-icon"
  >
    <StarFilledIcon
      customSize={16}
    />
  </div>
  <span>
    <span
      className="fdsChip-label"
    >
      foo
    </span>
    <span
      className="fdsChip-subtitle"
    >
      bar
    </span>
  </span>
  <span
    className="fdsChip-close border--focus"
    onClick={[Function]}
    role="button"
    tabIndex={0}
  >
    <DenyIcon
      customSize={10}
    />
  </span>
</button>
`;
